### PR TITLE
[Model] Add MGNNI and fix a bug in wikipedia_network

### PR DESCRIPTION
**Pull request by Eliaukyh**
…pedia_network中edge_index只有一个维度的问题。

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by Eliaukyh** [comment]
1）目前模型性能已经在cornell、texas和wisconsin三个数据集上达到要求；
2）修复了Wikipedia_network数据集中edge_index只有一个维度的bug；
3）暂时不支持mindspore框架，因为遇到了一些奇怪的bug，等过段时间调通了再提交。

**Comment by dddg617** [comment]
- Please remove the acc.csv file as we can put the outcome in the readme.md.
- What is the process.ipynb doing? If it is of no use, please remove this file.

**Comment by Eliaukyh** [comment]
Thank you for your review and feedback. I greatly appreciate your input. Based on your suggestions, I have taken the following actions:

* I have removed the acc.csv file and included the experimental results in a more organized format in the readme.md file. This will provide a clearer and more convenient way to access the outcomes.

* I have determined that the process.ipynb file is no longer necessary and have therefore removed it from the project.

Thank you for your valuable suggestions and guidance.

File: examples/mgnni/mgnni_trainner.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+# os.environ['TL_BACKEND'] = 'mindspore'
+# os.environ['TL_BACKEND'] = 'paddle'
+# os.environ['TL_BACKEND'] = 'tensorflow'  # set your backend here, default `tensorflow`
+os.environ["CUDA_VISIBLE_DEVICES"] = "0"
+import sys
+
+sys.path.insert(0, os.path.abspath('../../'))  # adds path2gammagl to execute in command line.
+import argparse
+import tensorlayerx as tlx  # 导入tensorlayerx包
+from gammagl.datasets import WebKB, WikipediaNetwork  # 专门做读取数据的类Planetoid
```
Comment (dddg617) [inline examples/mgnni/mgnni_trainner.py:17]: Remember to replace Chinese comment with English.

File: examples/mgnni/mgnni_trainner.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        dataset = WebKB(args.dataset_path, args.dataset, transform=T.NormalizeFeatures())
+    elif str.lower(args.dataset) in ['chameleon', 'squirrel']:
+        dataset = WikipediaNetwork(args.dataset_path, args.dataset, transform=T.NormalizeFeatures())
+    # dataset.process()
+    graph = dataset[0]
+    if args.dataset in ['cornell', 'texas', 'wisconsin']:
+        edge_index, _ = add_self_loops(graph.edge_index, num_nodes=graph.num_nodes, n_loops=args.self_loops)
+    else:
+        edge_index = graph.edge_index
+    edge_weight = tlx.convert_to_tensor(calc_gcn_norm(edge_index, graph.num_nodes))
```
Comment (dddg617) [inline examples/mgnni/mgnni_trainner.py:69]: Is this necessary to use ``convert_to_tensor``? I think the edge\_index is already tensor.

File: gammagl/layers/conv/mgnni_m_iter.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        self.f_solver = fwd_solver
+        self.b_solver = fwd_solver
+
+    def reset_parameters(self):
+        initor = tlx.initializers.Zeros()
+        self.F = self._get_weights("F", shape=self.F.shape, init=initor)
+
+    def _inner_func(self, Z, X, edge_index, edge_weight, num_nodes):
+        P = tlx.ops.transpose(Z)
+        ei = tlx.ops.convert_to_tensor(tlx.ops.convert_to_numpy(edge_index))
```
Comment (dddg617) [inline gammagl/layers/conv/mgnni_m_iter.py:27]: Why do we need ``convert_to_numpy`` and then use ``convert_to_tensor``? I think this is unnecessary.

File: gammagl/layers/conv/mgnni_m_iter.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        Z_new = self.gamma * g(self.F) @ Z + X
+        del Z, P, ei
+        return Z_new
+
+    def forward(self, X, edge_index, edge_weight, num_nodes):
+        Z, abs_diff = self.f_solver(lambda Z: self._inner_func(Z, X, edge_index, edge_weight, num_nodes),
+                                    z_init=tlx.zeros_like(X),
+                                    threshold=self.threshold,
+                                    max_iter=self.max_iter)
+        Z = tlx.convert_to_tensor(tlx.convert_to_numpy(Z))
```
Comment (dddg617) [inline gammagl/layers/conv/mgnni_m_iter.py:56]: Ditto, I do not think this is necessary.

File: gammagl/models/mgnni.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        self.fc2 = tlx.layers.Linear(out_features=nhid,
+                                     in_features=nhid,
+                                     # W_init='xavier_uniform',
+                                     )
+
+        self.dropout = dropout
+        self.MGNNI_layer = eval(fp_layer)(nhid, m_y, ks, threshold, max_iter, gamma, dropout=self.dropout,
+                                          batch_norm=batch_norm)
+
+    def forward(self, X, edge_index, edge_weight=None, num_nodes=None):
```
Comment (dddg617) [inline gammagl/models/mgnni.py:24]: Why not declare functions in the \_\_init\_\_? If you use like this way, there will be additional memory consumption when we need more layers as it will run the forward function many times.

File: gammagl/layers/conv/mgnni_m_iter.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
+import numpy as np
+import tensorlayerx as tlx
+import tensorlayerx.nn as nn
+
+from gammagl.utils import degree
+from gammagl.layers.conv import MessagePassing
+
+
+class MGNNI_m_iter(MessagePassing):
```
Comment (dddg617) [inline gammagl/layers/conv/mgnni_m_iter.py:9]: Remember to give a docstring refer to ``gcnconv``

File: gammagl/models/mgnni.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
+import numpy as np
+import tensorlayerx as tlx
+import tensorlayerx.nn as nn
+from gammagl.layers.conv import MGNNI_m_iter
+
+
+class MGNNI_m_MLP(tlx.nn.Module):
+    def __init__(self, m, m_y, nhid, ks, threshold, max_iter, gamma, fp_layer='MGNNI_m_att', dropout=0.5,
```
Comment (dddg617) [inline gammagl/models/mgnni.py:8]: Remember to give a description with docstring refer to model ``gcn``.

File: gammagl/models/mgnni.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        X = nn.Dropout(p=self.dropout)(tlx.ops.transpose(X))
+        X = nn.ReLU()(self.fc1(X))
+        X = nn.Dropout(p=self.dropout)(X)
+        X = self.fc2(X)
+        output = self.MGNNI_layer(tlx.ops.transpose(X), edge_index, edge_weight, num_nodes)
+
+        return output
+
+
+class MGNNI_m_att(nn.Module):
```
Comment (dddg617) [inline gammagl/models/mgnni.py:35]: Remember to give a description with docstring refer to model ``gcn``.

File: examples/mgnni/mgnni_trainner.py (hunk 1 above)
Comment (Eliaukyh) [reply examples/mgnni/mgnni_trainner.py]: Thank you for your reminder. I apologize for the oversight in leaving Chinese comments in the code. I will make sure to replace them with English comments to ensure clarity and consistency.

File: gammagl/layers/conv/mgnni_m_iter.py (hunk 3 above)
Comment (Eliaukyh) [reply gammagl/layers/conv/mgnni_m_iter.py]: Thank you for your feedback. Initially, my intention was to remove the gradient of that tensor by converting it to numpy and then back to a tensor. However, after your reminder, I realized that using convert_to_tensor() directly would suffice. I have made the necessary revisions to the code accordingly.

File: gammagl/layers/conv/mgnni_m_iter.py (hunk 4 above)
Comment (Eliaukyh) [reply gammagl/layers/conv/mgnni_m_iter.py]: Thank you for your feedback, my response is the same as above.

File: gammagl/models/mgnni.py (hunk 5 above)
Comment (Eliaukyh) [reply gammagl/models/mgnni.py:43]: Thank you very much for your reminder. I have realized the issue and have now moved the instantiation of dropout to the __init__ function.

File: gammagl/layers/conv/mgnni_m_iter.py (hunk 6 above)
Comment (Eliaukyh) [reply gammagl/layers/conv/mgnni_m_iter.py:9]: Thank you for the feedback. I apologize for the oversight. I have added the relevant docstring.

File: gammagl/models/mgnni.py (hunk 8 above)
Comment (Eliaukyh) [reply gammagl/models/mgnni.py:52]: Thank you for the feedback. I apologize for the oversight. I have added the relevant docstring.
